### DISCUSSION TITLE
acceleration input to crazyflie

### DISCUSSION THREAD

**Discussion by adarsh2798** [bitcraze/discussions#1225] [Q&A]
I am working on a project, where I am modelling my crazyflie as a double integrator. So I need to control it by giving it acceleration input(in world frame).
I am using the python API ``cflib``. I am familiar with sending position and velcoity setpoints. In ``commnader`` class I see a ``send_full_state_setpoint`` function where I can indeed send acceleration setpoint, however how to adjust other parameters, because the function expects position,velocity,orientaion,angular rates also. Can i modify the commander class to include a new function  that only accepts acceleration as arguments?

Basically I want to do acceleration control of  crazyflie, so Is there any other way I can send command corresponding to a desired acceleration? Is there any other provision already availabe in ``cflib``?

**Comment by knmcguire** [comment]
Hi!

That might be a bit tough todo to only give it acceleration. Currently only the [Mellinger](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/controller_mellinger.c) and [Brescianini controller ](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/controller_brescianini.c) use acceleration as an input and that is always in conjunction with a desired position and/or velocity so I'm not sure if without it it would work

Anyway, you'll need to define a new setpoint commander type, like add one here: https://github.com/bitcraze/crazyflie-firmware/blob/f61da11d54b6d54c7fa746688e8e9be4edd73a29/src/modules/src/crtp_commander_generic.c#L65-L74

then you need to write a decoder for it as well. Then you can add a cflib function to it like for the full state:
https://github.com/bitcraze/crazyflie-lib-python/blob/3e84ceefb3659da2d306c76db078eb9c2ef1f6ef/cflib/crazyflie/commander.py#L161-L189

If you don't want to change anything in the firmware or cflib, i'm not sure if that is possible. You could  log the state (velocity,postion,altitude,rate) and copy those values in the fullstate function, but by the time it got back to the crazyflie, the state values would probably have changed in the mean time so that loop will be too slow, but who knows

**Reply by adarsh2798** [reply]
Hi, Thanks for the reply
if I make changes in firmware as you mentiond, that's enough to handle acceleration inputs? The PID would take care of the rest? I dont need to write my own controller to handle the acceleration values and convert them to propellors PWM?
I see different controllers like ``attitude_PID_controller``,``controller_bresscianini``,etc.  Which one of them will handle if i only send acceleration as setpoints and nothing else? Should I make any extra changes in firmware for this or only changes in ``crtp_commnader_generaic.c`` file is enough?

***Update:***
I made the following changes And nothing really happens when I give acceleration command, crazyflie just hovers:


The changes i made in firmware are:

In crazyflie-firmware/src/modules/src, in the crtp_commander_generic.c file, I have made some changes.
```
/* ---===== 1 - packetType_e enum =====--- */ 
enum packet_type {
  stopType          = 0,
  velocityWorldType = 1,
  zDistanceType     = 2,
  cppmEmuType       = 3,
  altHoldType       = 4,
  hoverType         = 5,
  fullStateType     = 6,
  positionType      = 7,
  accType           = 8,                                      //edited(extended)
};
```

As shown, I am making changes for the commander.py class in cflib, to handle acceleration setpoints, so I have added ``accType=8`` as a new packet.
Then I add decoder function:
```
// ################################ EDITED(EXTENDED) PART##################################
/* accDecoder
 * Set the Crazyflie acceleration in the world coordinate system
 */
struct accPacket_s {
  float ax;        // m/s^2 in the world frame of reference
  float ay;        // ...
  float az;        // ...

} __attribute__((packed));
static void accDecoder(setpoint_t *setpoint, uint8_t type, const void *data, size_t datalen)
{const struct accPacket_s *values = data;

  ASSERT(datalen == sizeof(struct accPacket_s));

  setpoint->mode.x = modeAbs; // Mode for X acceleration
  setpoint->acceleration.x = values->ax;

  setpoint->mode.y = modeAbs; // Mode for Y acceleration
  setpoint->acceleration.y = values->ay;

  setpoint->mode.z = modeAbs; // Mode for Z acceleration
  setpoint->acceleration.z = values->az;
}

//#############################################################################################
```

then finally in packetDecoder array i add `` [accType] = accDecoder``
```
 /* ---===== 3 - packetDecoders array =====--- */
const static packetDecoder_t packetDecoders[] = {
  [stopType]          = stopDecoder,
  [velocityWorldType] = velocityDecoder,
  [zDistanceType]     = zDistanceDecoder,
  [cppmEmuType]       = cppmEmuDecoder,
  [altHoldType]       = altHoldDecoder,
  [hoverType]         = hoverDecoder,
  [fullStateType]     = fullStateDecoder,
  [positionType]      = positionDecoder,
  [accType]           = accDecoder,//edited(extended)
};
```

These are the only changes i made.

Also I added another functio in commander class: (TYPE_ACCEL_WORLD=8 as defined in enum packets)

```
 def send_accel_world_setpoint(self,ax,ay,az):
    
        """
        Send acceleration in the world frame of reference 

        ax, ay, az are in m/s^2
    
        """
        pk = CRTPPacket()
        pk.port = CRTPPort.COMMANDER_GENERIC
        pk.channel = SET_SETPOINT_CHANNEL
        pk.data = struct.pack('<Bfff', TYPE_ACCEL_WORLD,
                              ax, ay, az)
        self._cf.send_packet(pk)
```

**Reply by knmcguire** [reply]
hi! So indeed it seems that you have found the right steps to get the  message across, but no the PID controller will not automatically handle acceleration as it doesn't do anything with acceleration

Also the following:

```
  setpoint->mode.x = modeAbs; // Mode for X acceleration
  setpoint->acceleration.x = values->ax;
```


modeAbs is the absolute mode will only tell the controllers (mostly PID controller), to control the position. There is currently no setpoint set for position so that is perhaps initialized at zero I believe? 

And you should try out the mellinger or bresscianini controller if you see any response. 


But also mind that none of us has tried it before either so we don't have clue if this would actually work. If this doesn't work on the two controllers I've mentioned then I'm afraid you'll need to write a controller in the firmware yourself.
